src/memtable.rs

```diff
@@ -675,52 +742,96 @@ impl Drop for MemTable {
     /// Deleted [`MemTable`]s that are not yet rewritten.
     removed_memtables: Arc<Mutex<VecDeque<u64>>>,
 }
 
-impl MemTableAccessor {
-    pub fn new(global_stats: Arc<GlobalStats>) -> MemTableAccessor {
+impl MemTableAccessor<VacantAllocator> {
+    pub fn new(global_stats: Arc<GlobalStats>) -> MemTableAccessor<VacantAllocator> {
+        let mut slots = Vec::with_capacity(MEMTABLE_SLOT_COUNT);
+        for _ in 0..MEMTABLE_SLOT_COUNT {
+            slots.push(Arc::new(RwLock::new(MemTableMap::default())));
+        }
+        MemTableAccessor {
+            global_stats,
+            allocator: new_vacant_allocator(),
+            slots,
+            removed_memtables: Default::default(),
+        }
+    }
+}
+
+impl MemTableAccessor<SelectedAllocator> {
+    pub fn memory_usage(&self) -> usize {
+        #[cfg(not(feature = "swap"))]
+        {
+            let mut total = 0;
+            for tables in &self.slots {
+                tables.read().values().for_each(|t| {
+                    total += t.read().heap_size();
+                });
+            }
+            total
+        }
+        #[cfg(feature = "swap")]
+        {
+            self.allocator.memory_usage()
+        }
+    }
+
+    pub(crate) fn flush_metrics(&self) {
+        MEMORY_USAGE.set(self.memory_usage() as i64);
+    }
+}
+
+impl<A: AllocatorTrait> MemTableAccessor<A> {
+    pub fn new_with_allocator(global_stats: Arc<GlobalStats>, allocator: A) -> MemTableAccessor<A> {
         let mut slots = Vec::with_capacity(MEMTABLE_SLOT_COUNT);
         for _ in 0..MEMTABLE_SLOT_COUNT {
-            slots.push(Arc::new(RwLock::new(MemTables::default())));
+            slots.push(Arc::new(RwLock::new(MemTableMap::default())));
         }
         MemTableAccessor {
             global_stats,
+            allocator,
             slots,
             removed_memtables: Default::default(),
         }
     }
 
-    pub fn get_or_insert(&self, raft_group_id: u64) -> Arc<RwLock<MemTable>> {
+    pub fn get_or_insert(&self, raft_group_id: u64) -> Arc<RwLock<MemTable<A>>> {
         let global_stats = self.global_stats.clone();
         let mut memtables = self.slots[Self::slot_index(raft_group_id)].write();
-        let memtable = memtables
-            .entry(raft_group_id)
-            .or_insert_with(|| Arc::new(RwLock::new(MemTable::new(raft_group_id, global_stats))));
+        let memtable = memtables.entry(raft_group_id).or_insert_with(|| {
+            let memtable =
+                MemTable::with_allocator(raft_group_id, global_stats.clone(), &self.allocator);
+            Arc::new(RwLock::new(memtable))
+        });
         memtable.clone()
     }
 
-    pub fn get(&self, raft_group_id: u64) -> Option<Arc<RwLock<MemTable>>> {
+    pub fn get(&self, raft_group_id: u64) -> Option<Arc<RwLock<MemTable<A>>>> {
         self.slots[Self::slot_index(raft_group_id)]
             .read()
             .get(&raft_group_id)
             .cloned()
```

Comment:
can we return reference?

Reply:
Afraid not. The reference requires the lock on `slots`.